### PR TITLE
feat: add acert CLI

### PR DESCRIPTION
**Pull request by JonathanWilbur**
This PR adds support for the `acert` subcommand, which allows printing, signing, and verifying attribute certificates.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by JonathanWilbur** [comment]
@dhobsong Just letting you know about this.

File: apps/acert.c [hunk 1]
```diff
@@ -0,0 +1,840 @@
+/*
```
Comment (paulidale) [inline apps/acert.c:1]: This doesn't seem to be linked into the openssl command line app.
Likewise, the docs seem absent.
Tests are also missing.

**Comment by t-j-h** [comment]
I don't think these are a separate utility program "acert" - these are basically all to do with handling extensions within the context of x509 certificates - so why wouldn't they be in the "x509" program. We have some level of extension handling there and for this context fleshing that out to be a better set of options for command line users would make sense.

Others?

**Comment by t8m** [comment]
@t-j-h Please note this is about managing X509_ACERT certificates not regular X.509 certificates. They are not identical.

**Comment by beldmit** [comment]
How many options for x509 utility are irrelevant for acert utility?

**Comment by t-j-h** [comment]
> @t-j-h Please note this is about managing X509_ACERT certificates not regular X.509 certificates. They are not identical.

At least it should be called x509_acert then ... not acert ... and a *lot* of the x509 app functionality will end up having to be duplicated.

**Comment by t8m** [comment]
Maybe it would make sense to instead add an `-acert` or `-attrcert` option to the x509 command which would make it to handle attribute certificates instead of regular ones?

**Comment by t-j-h** [comment]
Separately, those parts of the PR adding new acert APIs should be in a separate PR and reviewed for consistency with the existing APIs and the X509 APIs - we shouldn't have those combined with the addition of an app.

And a lot of testing stuff needs to go in ... which should be on the APIs themselves  IMHO.

**Comment by dhobsong** [comment]
Hi @JonathanWilbur, thanks for letting me know.

I also have to echo the comment above that the addition of new APIs should be handled in a separate PR (or at the very least a separate commit) from adding the app.

**Comment by JonathanWilbur** [comment]
I will opine on a few things:

1. I think attribute certificate signing, verification, and creation should take place in its own subcommand for no other reason than complexity: this subcommand is going to have a lot of options that don't overlap with X.509 public key certificates. This could be even more pronounced if we ever add more features for adding attributes via the CLI. In theory, it is possible to just verify an attribute certificate along with a PKC chain, but IMO, it makes sense to adhere to the Unix philosophy of one command that does only one thing and does it well (that "one thing" already being three things in this case).
2. Even I, as the PR submitter, agree that this should be broken up into smaller PRs. I will close this out and do that.

**Comment by t-j-h** [comment]
@JonathanWilbur the contributions are very much welcomed and we are working to improve processes for contributor input.

On new APIs - it makes sense to outline the API and get that reviewed separate from the implementation if you want to avoid some churn - as we look for consistency in APIs and ease of use. There are some odd quirks in the API set which need adjusting and some more documentation around the APIs would help too.

If you can separate things out that way it helps - but if you want to have the implementation and the APIs in one and work through changing names and parameters around on-the-fly that can also be done - but it does tend to lead to more effort.

Then testing of the APIs. And some test data.

Then the proposed application command line options - compared to the x509 command - what is in common between x509-acert/x509acert and x509 - and what is distinct for each.

Then the implementation of the application can be done separately - again trying to save on churn.

In writing the application you may see that helper functions make using the APIs better - those should be considered for addition - so we don't have every application needing to cut-n-paste large parts of application code to do useful things.

The larger the chunk of code that comes in at once is, the more challenging it is for the reviewers - and also the larger impact any feedback has on the contributor needing to change things based on feedback - taking the approach in multiple steps should get faster feedback and involve less churn and less work for everyone involved to hit your objective of getting a nice stable, long term supportable API for attribute certificate usage into OpenSSL.

Hopefully the list above helps.

**Comment by beldmit** [comment]
Shouldn't a feature branch be used for this efforts?

**Comment by JonathanWilbur** [comment]
Thanks for the feedback @t-j-h , that sounds like a fine plan. I am curious what you mean by "outline the API" though. Are you suggesting that I define a header file containing the function declarations of what I propose, and to add it to OpenSSL in such a way where it is not built or used anywhere (at least until implementation)?

More generally, what I am asking is: how would you like me to implement these non-functional, granular, incremental PRs, such as just the API outline and just the command-line parameters alone? To what @beldmit is saying, it sounds like a feature branch might be nice for this purpose.

**Comment by t-j-h** [comment]
Header file for the API outline - we don't need the implementation to provide feedback on the API. It also helps to have the documentation for the API as well.

**Comment by t-j-h** [comment]
@t8m do you think a feature branch for this work makes sense?

**Comment by t8m** [comment]
Perhaps, if the command line app will need to add a lot of new APIs then yes.

**Comment by mattcaswell** [comment]
+1 for a feature branch for this.

@JonathanWilbur can you please request one as per:

https://openssl-library.org/policies/general/feature-branch-approval-policy/

**Comment by JonathanWilbur** [comment]
Sure thing! To address @t8m 's question, I don't anticipate this API being large. Namely, I can think of:

- `sign()`
- `verify()`
- `print()`
- `check_crl()` (questionable, since you could probably just extract the serial number and use existing functions
- `check_ocsp()` (Again, this might be able to re-use existing functions)

If it isn't considered bad practice, there may be `_ex` variants of the above, but the last two don't really have to be done in this PR at all, IMO. I don't think we will have to modify existing important structures like `X509_STORE_CTX`, because I wasn't planning on this doing full X.509 PMI verification (e.g. an "attribute certificate chain"); all I was planning on this doing was verifying that a single attribute certificate is valid with respect to a public-key holder and attribute authority.

**Comment by JonathanWilbur** [comment]
To clarify: what I said does not preclude also verifying the extensions, `notBefore` and `notAfter` times, etc. All of the things you would expect to be verified will be verified.

File: include/openssl/x509_acert.h.in [hunk 2]
```diff
@@ -196,4 +200,24 @@ DECLARE_ASN1_FUNCTIONS(OSSL_AUTHORITY_ATTRIBUTE_ID_SYNTAX)
     generate_stack_macros("OSSL_ISSUER_SERIAL");
 -}
 
+int X509V3_EXT_ACERT_add_nconf(CONF *conf, X509V3_CTX *ctx, const char *section,
+                               X509_ACERT *acert);
+int X509V3_EXT_ACERT_add_conf(LHASH_OF(CONF_VALUE) *conf, X509V3_CTX *ctx,
+                              const char *section, X509_ACERT *acert);
+
+int ossl_x509_check_acert_time(X509_STORE_CTX *ctx, X509_ACERT *acert);
+int ossl_x509_check_acert_exts(X509_ACERT *acert);
```
[the remark this answers is not in the document]
Comment (JonathanWilbur) [reply include/openssl/x509_acert.h.in:209]: Yep, that looks like a mistake on my part. I'll change it.

**Comment by JonathanWilbur** [comment]
Actually, I take that back. I know I am outnumbered here, but I don't think this should be in a feature branch at all. This PR (which is not really that big overall), can be split into several smaller PRs already, each of which are individually useful, _and_ they could be merged in any any order. I could create a PR for a CLI that supports no signing and verification and only printing, or I could create a PR that supports signing and verifying attribute certificates with no CLI, and even these can be broken down further into useful atoms.

The overhead of repeatedly having to rebase and the possibility of these changes not making it into 3.5.0 because they are collectively only 90% done does not sound worth it to me.

But that is just a point made about the technical usefulness of a feature branch. If you still want a feature branch as an administrative formality, I will still do whatever you want.

**Comment by JonathanWilbur** [comment]
Submitted a new feature branch request: https://github.com/openssl/project/issues/845

**Comment by t-j-h** [comment]
A feature brach is the right thing as all of these are related and should either all go in or none go in. There are tests and docs and API and util approaches all which need review on consistency with the existing approaches.

**Comment by DDvO** [comment]
I just became aware of this PR.

> I don't think these are a separate utility program "acert" - these are basically all to do with handling extensions within the context of x509 certificates - so why wouldn't they be in the "x509" program. We have some level of extension handling there and for this context fleshing that out to be a better set of options for command line users would make sense.
> 
> Others?

I fully agree, for two reasons:
* While most people are not aware, also attribute certs are X.509 certs.
The "regular X.509 certificates", as @t8m called them above, should more precisely called "X.509 public-key certificates".
* Adding a separate CLI app for X.509 attribute certs would just increase the code redundancy mess in `apps/`.
Instead, better tidy up `apps/x509.c` and then extend it with specific acert options.

**Comment by esyr** [comment]
> Shouldn't a feature branch be used for this efforts?

I concur.

**Pull request closed without merging**